url/nodeurl.go

```diff
@@ -0,0 +1,117 @@
+package url
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type searchParam struct {
+	name  string
+	value string
+}
+
+func (sp *searchParam) Encode() string {
+	return sp.string(true)
+}
+
+func (sp *searchParam) string(encode bool) string {
+	vals := []string{}
+	if encode {
+		vals = append(vals, fmt.Sprintf("%s=%s", url.QueryEscape(sp.name), url.QueryEscape(sp.value)))
+	} else {
+		vals = append(vals, fmt.Sprintf("%s=%s", sp.name, sp.value))
+	}
+	return strings.Join(vals, "&")
+}
+
+type searchParams []searchParam
+
+func (s searchParams) Len() int {
+	return len(s)
+}
+
+func (s searchParams) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s searchParams) Less(i, j int) bool {
+	return len(s[i].name) > len(s[j].name)
+}
+
+func (s searchParams) Encode() string {
+	str := ""
+	sep := ""
+	for _, v := range s {
+		str = fmt.Sprintf("%s%s%s", str, sep, v.Encode())
+		sep = "&"
+	}
+	return str
+}
+
+func (s searchParams) String() string {
+	var b strings.Builder
+	sep := ""
+	for _, v := range s {
+		b.WriteString(sep)
+		b.WriteString(v.string(false)) // keep it raw
+		sep = "&"
+	}
+	return b.String()
+}
+
+type nodeURL struct {
+	url          *url.URL
+	searchParams searchParams
+}
+
+// This methods ensures that the url.URL has the proper RawQuery based on the searchParam
+// structs. If a change is made to the searchParams we need to keep them in sync.
+func (nu *nodeURL) syncSearchParams() {
+	nu.url.RawQuery = nu.searchParams.Encode()
+}
+
+func (nu *nodeURL) String() string {
+	return nu.url.String()
+}
+
+func (nu *nodeURL) hasName(name string) bool {
+	for _, v := range nu.searchParams {
+		if v.name == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (nu *nodeURL) getValues(name string) []string {
+	vals := []string{}
```

Comment:
Interesting, the reason I had done this was to avoid a crash when we called `len(vals)` where the value is returned. I didn't know that `len(x)` could take nil as a value. Good catch!